### PR TITLE
Move past active deadline check into syncLoop of kubelet.

### PR DESCRIPTION
**Pull request by Random-Liu**
Move past active deadline check into syncLoop of kubelet. The pod worker will be waked up only when kubelet find it passed the active deadline.

(For #12540)
/cc @yujuhong

### CODE REVIEW

**Comment by Random-Liu** [comment]
@k8s-bot unit test this.

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -1776,6 +1779,50 @@ func (kl *Kubelet) pastActiveDeadline(pod *api.Pod) bool {
...
+	// Get all pods whose work are ready
+	podUIDs := kl.workQueue.GetWork()
+	var podsToWork []*api.Pod
+	for _, uid := range podUIDs {
+		if pod, ok := kl.podManager.GetPodByUID(uid); ok {
+			podsToWork = append(podsToWork, pod)
+		}
+	}
+
+	// Merge the pod lists
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: The merging code is abit hard to expand if we add more criteria along the way. My suggestion would be:
 1. Get all the pods at the beginning of this function.
 2. Get the podUIDs from the work queue, and create a set of uids.
 3. Iterate through every pod from (1), and check whether they matches any of the following criteria; append if it's true
    - pastActiveDeadline == true
    - uid in the set from (2)
4. Return the list from (3)

Comment (Random-Liu) [inline pkg/kubelet/kubelet.go]: Done! Your suggestion makes the code much simpler! Thanks a lot! :)

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -1764,6 +1767,32 @@ func (kl *Kubelet) pastActiveDeadline(pod *api.Pod) bool {
...
 }
 
+// Get pods which should be resynchronized. Move the logic here so as to make it testable and make the main sync loop
+// simple and clear.
+// Currently, the following pod should be resynchronized:
+//   * pod whose work is ready.
+//   * pod past the active deadline.
+func (kl *Kubelet) getPodsToSync() []*api.Pod {
+	allPods := kl.podManager.GetPods()
+	podUIDs := kl.workQueue.GetWork()
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: Might be easier to build a set out of the list, e.g., `util.set.NewString()`, then check whether a pod uid is in the set without having to use a for loop.

File: pkg/kubelet/kubelet.go [hunk 3]
```diff
@@ -1764,6 +1767,32 @@ func (kl *Kubelet) pastActiveDeadline(pod *api.Pod) bool {
 	return false
 }
 
+// Get pods which should be resynchronized. Move the logic here so as to make it testable and make the main sync loop
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: No need to write a comment like this if it's obvious :)

File: pkg/kubelet/kubelet.go (hunk 2 above)
Comment (Random-Liu) [inline pkg/kubelet/kubelet.go]: Yeah, that's more efficient :) I'll do it!

File: pkg/kubelet/kubelet.go (hunk 3 above)
Comment (Random-Liu) [inline pkg/kubelet/kubelet.go]: Done :)

File: pkg/kubelet/kubelet.go [hunk 4]
```diff
@@ -1764,6 +1767,31 @@ func (kl *Kubelet) pastActiveDeadline(pod *api.Pod) bool {
...
+	podUIDs := kl.workQueue.GetWork()
+	podUIDSet := sets.NewString()
+	for _, podUID := range podUIDs {
+		podUIDSet.Insert(string(podUID))
+	}
+	var podsToSync []*api.Pod
+	for _, pod := range allPods {
+		if kl.pastActiveDeadline(pod) {
+			// The pod has passed active deadline
+			podsToSync = append(podsToSync, pod)
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: minor nit: add `continue` here and after line 1788. This removes the else block and allows you to add more `if` in the future.

Comment (Random-Liu) [inline pkg/kubelet/kubelet.go]: Wow! Got it!

**Comment by yujuhong** [comment]
lgtm. please squash. thanks!

**Comment by Random-Liu** [comment]
Sure, :)

**Comment by Random-Liu** [comment]
Squashed, :)

**Comment by Random-Liu** [comment]
@k8s-bot e2e test this.

**Comment by Random-Liu** [comment]
@k8s-bot e2e test this.

**Pull request merged**